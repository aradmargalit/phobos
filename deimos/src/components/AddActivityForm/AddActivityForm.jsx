import './AddActivityForm.scss';

import { EditOutlined, RocketOutlined } from '@ant-design/icons';
import {
  Button,
  DatePicker,
  Form,
  Input,
  InputNumber,
  message,
  notification,
  Select,
  Spin,
} from 'antd';
import moment from 'moment';
import React, { useContext, useState } from 'react';

import { postActivity, putActivity } from '../../apis/phobos-api';
import UserContext from '../../contexts';
import CalculatedActivityFields from '../CalculatedActivityFields';
import EmojiOption from '../EmojiOption';


const dateFormat = 'MMMM Do, YYYY';

const { Item } = Form;
const { Option } = Select;

export default function AddActivityForm({
  activityTypes, loading, refetch, initialActivity, modalClose,
}) {
  const { user } = useContext(UserContext);
  const [activity, setActivity] = useState({ ...initialActivity, unit: 'miles' });
  const [form] = Form.useForm();

  const editing = !!initialActivity;

  const upsert = async (values, apiCall) => {
    try {
      await apiCall(values);
      message.success(`Successfully ${editing ? 'updated' : 'created'} activity!`);
      refetch();
    } catch (err) {
      notification.error({
        message: 'Unexpected Error',
        description: `Error: ${err}`,
      });
    }
  };

  const onFinish = async (values) => {
    if (editing) {
      const activityToPut = { ...values };
      activityToPut.id = initialActivity.id;
      await upsert(activityToPut, putActivity);
      modalClose();
    } else {
      await upsert(values, postActivity);
    }
  };

  const onReset = () => {
    form.resetFields();
    setActivity(form.getFieldsValue());
  };

  const onSubmit = () => {
    form.submit();
  };

  const onChange = (_, all) => {
    setActivity(all);
  };

  const layout = {
    labelCol: {
      span: 6,
    },
    wrapperCol: {
      span: 12,
    },
  };

  return (
    <Spin spinning={loading}>
      <div className="form-flex">
        <Form
          {...layout}
          form={form}
          name="add-activity"
          onFinish={onFinish}
          onValuesChange={onChange}
          initialValues={{ activity_date: moment(new Date()), unit: 'miles', ...initialActivity }}
        >
          {/* ============= NAME ============= */}
          <Item
            hasFeedback
            label="Activity Name"
            name="name"
          >
            <Input className="fullWidth" placeholder={`${user.given_name}'s Favorite Run`} />
          </Item>
          {/* ============= DATEPICKER ============= */}
          <Item
            hasFeedback
            label="Activity Date"
            name="activity_date"
            rules={[
              {
                required: true,
                message: 'Activity Date is required',
              },
            ]}
          >
            <DatePicker format={dateFormat} className="fullWidth" placeholder="2020-01-01" />
          </Item>

          {/* ============= ACTIVITY SELECT ============= */}
          <Item
            hasFeedback
            label="Activity Type"
            name="activity_type_id"
            rules={[
              {
                required: true,
                message: 'Activity Type is required',
              },
            ]}
          >
            <Select allowClear showSearch optionFilterProp="children">
              {activityTypes.map(
                ({ id, name }) => EmojiOption({ value: id, title: name }),
              )}
            </Select>
          </Item>

          {/* ============= DURATION ============= */}
          <Item
            hasFeedback
            label="Duration (min)"
            name="duration"
            rules={[{ required: true, message: 'Duration is required' }]}
          >
            <InputNumber
              className="fullWidth"
              precision={2}
              placeholder={30}
              min={0.5}
              step={0.5}
            />
          </Item>

          {/* ============= DISTANCE ============= */}
          <Item label="Distance" style={{ marginBottom: 0 }}>
            <Item name="distance" className="inline-item">
              <InputNumber
                precision={2}
                min={0.1}
                step={0.1}
                placeholder={5}
              />
            </Item>
            <Item name="unit" className="inline-item">
              <Select>
                <Option value="miles">miles</Option>
                <Option value="yards">yards</Option>
              </Select>
            </Item>
          </Item>
        </Form>
        <CalculatedActivityFields activity={activity} />
      </div>
      <div className="button-row">
        <Button className="button-row-item" onClick={onSubmit} icon={editing ? <EditOutlined /> : <RocketOutlined rotate={45} />} type="primary">
          {editing ? 'Edit' : 'Submit'}
        </Button>
        <Button className="button-row-item" ghost onClick={onReset} type="primary">
          Reset
        </Button>
      </div>
    </Spin>
  );
}
