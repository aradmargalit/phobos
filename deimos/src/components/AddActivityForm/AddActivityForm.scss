.emoji {
  padding-left: 10px;
}

.fullWidth {
  width: 100%;
}

.inline-item {
  display: inline-block;
  width: calc(50% - 6px);
}

.ant-form {
  width: 80%;
  text-align: left;
}

.ant-statistic {
  width: 15%;
}

.form-flex {
  display: flex;
}

.button-row {
  margin: 15px 0px;
  display: flex;
  justify-content: center;

  &-item {
    margin-left: 3px;
    width: 100px;
  }
}

.unit {
  margin-left: auto;
}

.distance-wrapper > .ant-form-item-control > .ant-form-item-control-input {
  display: flex;

  & > .ant-form-item {
    margin-bottom: 0;
  }
}