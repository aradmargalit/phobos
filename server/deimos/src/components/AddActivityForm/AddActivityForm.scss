.emoji {
  padding-left: 10px;
}

.fullWidth {
  width: 100%;
}

.inline-item {
  display: inline-block;
  width: calc(50% - 6px);
}

.ant-form {
  width: 80%;
  text-align: left;
}

.form-statistic {
  width: 15%;
}

.form-flex {
  display: flex;
}

.button-row {
  margin: 15px 0px;
  display: flex;
  justify-content: center;

  &-item {
    margin-left: 3px;
    width: 100px;
  }
}

.unit {
  margin-left: auto;
}

.distance-wrapper > .ant-form-item-control > .ant-form-item-control-input {
  display: flex;

  & > .ant-form-item {
    margin-bottom: 0;
  }
}

.ant-spin-container {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 90%;
}

.ant-spin-nested-loading {
  height: 100%;
}
