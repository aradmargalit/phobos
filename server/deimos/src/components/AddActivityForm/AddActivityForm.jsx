import './AddActivityForm.scss';

import { DatePicker, Form, Input, InputNumber, Select } from 'antd';
import moment from 'moment';
import React from 'react';

import DurationPicker from '../DurationPicker';
import EmojiOption from '../EmojiOption';

const dateFormat = 'MMMM Do, YYYY';

const { Item } = Form;
const { Option } = Select;

export default function AddActivityForm({
  form,
  user,
  onFinish,
  activityTypes,
  setActivity,
  initialActivity,
}) {
  const onChange = (_, all) => {
    setActivity(all);
  };

  const layout = {
    labelCol: {
      span: 6,
    },
    wrapperCol: {
      span: 12,
    },
  };

  return (
    <div className="form-flex">
      <Form
        {...layout}
        autoComplete="off"
        form={form}
        name="add-activity"
        onFinish={onFinish}
        onValuesChange={onChange}
        initialValues={{
          activity_date: moment(new Date()),
          unit: 'miles',
          ...initialActivity,
        }}
      >
        {/* ============= NAME ============= */}
        <Item label="Activity Name" name="name">
          {/* Data LP ignore is to stop LastPass from trying to fill the form */}
          <Input
            data-lpignore="true"
            className="fullWidth"
            placeholder={user && `${user.given_name}'s Favorite Run`}
          />
        </Item>
        {/* ============= DATEPICKER ============= */}
        <Item
          label="Activity Date"
          name="activity_date"
          rules={[
            {
              required: true,
              message: 'Activity Date is required',
            },
          ]}
        >
          <DatePicker
            format={dateFormat}
            className="fullWidth"
            placeholder="2020-01-01"
          />
        </Item>
        {/* ============= ACTIVITY SELECT ============= */}
        <Item
          label="Activity Type"
          name="activity_type_id"
          rules={[
            {
              required: true,
              message: 'Activity Type is required',
            },
          ]}
        >
          <Select
            allowClear
            showSearch
            placeholder="Run"
            optionFilterProp="children"
          >
            {activityTypes.map(({ id, name }) =>
              EmojiOption({ value: id, title: name })
            )}
          </Select>
        </Item>
        {/* ============= DURATION ============= */}
        <Item
          label="Duration"
          name="duration"
          rules={[{ required: true, message: 'Duration is required' }]}
        >
          <DurationPicker />
        </Item>
        {/* ============= DISTANCE ============= */}
        <Item label="Distance" className="distance-wrapper">
          <Item name="distance" className="inline-item">
            <InputNumber precision={2} min={0.1} step={0.1} placeholder={5} />
          </Item>
          <Item name="unit" className="unit inline-item">
            <Select>
              <Option value="miles">miles</Option>
              <Option value="yards">yards</Option>
            </Select>
          </Item>
        </Item>
      </Form>
    </div>
  );
}
