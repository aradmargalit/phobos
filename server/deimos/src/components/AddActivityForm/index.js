import AddActivityForm from './AddActivityForm';

export default AddActivityForm;
