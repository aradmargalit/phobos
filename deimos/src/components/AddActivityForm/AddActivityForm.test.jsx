import { shallow } from 'enzyme';
import React from 'react';

import AddActivityForm from './AddActivityForm';

const generateComponent = props => {
  const defaultProps = {
    user: { given_name: 'Test' },
    activityTypes: [
      { id: 1, name: 'Running' },
      { id: 2, name: 'Biking' },
    ],
  };
  const mergedProps = { ...defaultProps, ...props };

  return shallow(<AddActivityForm {...mergedProps} />);
};

describe('<AddActivityForm />', () => {
  it('renders correctly', () => {
    expect(generateComponent()).toMatchSnapshot();
  });
});
