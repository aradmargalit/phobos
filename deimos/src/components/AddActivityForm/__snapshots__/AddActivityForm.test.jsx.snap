// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddActivityForm /> renders correctly 1`] = `
<div
  className="form-flex"
>
  <ForwardRef(InternalForm)
    autoComplete="off"
    initialValues={
      Object {
        "activity_date": "2020-06-15T02:13:54.115Z",
        "unit": "miles",
      }
    }
    labelCol={
      Object {
        "span": 6,
      }
    }
    name="add-activity"
    onValuesChange={[Function]}
    wrapperCol={
      Object {
        "span": 12,
      }
    }
  >
    <FormItem
      label="Activity Name"
      name="name"
    >
      <Input
        className="fullWidth"
        data-lpignore="true"
        placeholder="Example: Test's Favorite Run"
        type="text"
      />
    </FormItem>
    <FormItem
      label="Activity Date"
      name="activity_date"
      rules={
        Array [
          Object {
            "message": "Activity Date is required",
            "required": true,
          },
        ]
      }
    >
      <Picker
        className="fullWidth"
        format="MMMM Do, YYYY"
        placeholder="2020-01-01"
      />
    </FormItem>
    <FormItem
      label="Activity Type"
      name="activity_type_id"
      rules={
        Array [
          Object {
            "message": "Activity Type is required",
            "required": true,
          },
        ]
      }
    >
      <Select
        allowClear={true}
        bordered={true}
        choiceTransitionName="zoom"
        optionFilterProp="children"
        placeholder="Choose your activity type..."
        showSearch={true}
        transitionName="slide-up"
      >
        <Option
          key="1"
          value={1}
        >
          Running
        </Option>
        <Option
          key="2"
          value={2}
        >
          Biking
        </Option>
      </Select>
    </FormItem>
    <FormItem
      label="Duration"
      name="duration"
      rules={
        Array [
          Object {
            "message": "Duration is required",
            "required": true,
          },
        ]
      }
    >
      <DurationPicker />
    </FormItem>
    <FormItem
      className="distance-wrapper"
      label="Distance"
    >
      <FormItem
        className="inline-item"
        name="distance"
      >
        <InputNumber
          min={0}
          placeholder={5}
          precision={2}
          step={0.1}
        />
      </FormItem>
      <FormItem
        className="unit inline-item"
        name="unit"
      >
        <Select
          bordered={true}
          choiceTransitionName="zoom"
          transitionName="slide-up"
        >
          <Option
            value="miles"
          >
            miles
          </Option>
          <Option
            value="yards"
          >
            yards
          </Option>
        </Select>
      </FormItem>
    </FormItem>
    <FormItem
      label="Heart Rate"
      name="heart_rate"
    >
      <InputNumber
        className="fullWidth"
        data-lpignore="true"
        max={250}
        min={1}
        placeholder="Average Beats per Minute"
        step={1}
      />
    </FormItem>
  </ForwardRef(InternalForm)>
</div>
`;
